scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,391 @@ def recv(self, x=0xffff):
+    """ Craft ISO TP packet
+    Args:
+        identifier: identifier of crafted packet
+        extended: boolean if packet uses extended address
+    """
+
+    if extended:
+        pkt = ISOTPHeaderEA() / ISOTP_FF()
+        pkt.extended_address = 0
+        pkt.data = b'\x00\x00\x00\x00\x00'
+    else:
+        pkt = ISOTPHeader() / ISOTP_FF()
+        pkt.data = b'\x00\x00\x00\x00\x00\x00'
+
+    pkt.identifier = identifier
+    pkt.message_size = 100
+    return pkt
+
+
+def filter_periodic_packets(packet_list, verbose=False):
+    """ Filter for periodic packets
+
+    Args:
+        packet_list: Dictionary with Send-to-ID as key and a tuple
+        (received packet, Recv_ID)
+        verbose: Displays further information
+
+    ISOTP-Filter for periodic packets (same ID, always same timegap)
+    Deletes periodic packets in packet_list
+    """
+
+    delete_list = set()
+    filter_dict = {}
+
+    for _, value in packet_list.items():
+        pkt = value[0]
+        idn = value[1]
+        if idn not in filter_dict.keys():
+            filter_dict[idn] = [pkt]
+        else:
+            filter_dict[idn] += [pkt]
+
+    for idn in filter_dict:
+        pkt_lst = filter_dict[idn]
+        if len(pkt_lst) < 3:
+            continue
+
+        tg = [p1.time - p2.time for p1, p2 in zip(pkt_lst[1:], pkt_lst[:-1])]
+        if all(abs(t1 - t2) < 0.001 for t1, t2 in zip(tg[1:], tg[:-1])):
+            if verbose:
+                print("[i] Identifier 0x%03x seems to be periodic. "
+                      "Filtered.")
+            delete_list.add(idn)
+
+    for key in packet_list.copy():
+        if packet_list[key][1] in delete_list:
+            del packet_list[key]
+
+
+def get_isotp_fc(id_value, id_list, noise_ids, extended, packet,
+                 verbose=False):
+    """Callback for sniff function when packet received
+
+    Args:
+            id_value: packet id of send packet
+            id_list: list of received IDs
+            noise_ids: list of packet IDs which will not be considered when
+                       received during scan
+            extended: boolean if extended scan
+            packet: received packet
+            verbose: displays information during scan
+
+    If received packet is a FlowControl
+    and not in noise_ids
+    append it to id_list
+    """
+    if packet.flags != 0:
+        return
+
+    if noise_ids is not None and packet.identifier in noise_ids:
+        return
+
+    try:
+        index = 1 if extended else 0
+        isotp_pci = orb(packet.data[index]) >> 4
+        isotp_fc = orb(packet.data[index]) & 0x0f
+        if isotp_pci == 3 and isotp_fc in [0, 1, 2]:
+            if verbose:
+                print("[+] Found flow-control frame from identifier 0x%03x" %
+                      packet.identifier)
+            if isinstance(id_list, dict):
+                id_list[id_value] = (packet, packet.identifier)
+            if isinstance(id_list, list):
```

Comment:
`elif`
Since it is no `dict` and `list` at the same time.

Reply:
done